### PR TITLE
Document when Camera::viewport_to_world and related methods return None

### LINKED ISSUE #7171: Document when `Camera::viewport_to_world` and related methods return `None`
## How can Bevy's documentation be improved?

Document when [`viewport_to_world`](https://docs.rs/bevy/latest/bevy/render/camera/struct.Camera.html#method.viewport_to_world) and related methods return `None`.

### PR DESCRIPTION
**Pull request by tormeh**
# Objective

Fixes #7171

## Solution

- Add documentation for when Camera::viewport_to_world and related methods return None

### CODE REVIEW

**Comment by tormeh** [comment]
This is my first Bevy PR and I'm a total rendering noob, so I hope I got it right :sweat_smile:

**Review by alice-i-cecile — APPROVED** [review]
Good! I'm not 100% convinced that this is exhaustive, but I couldn't find any other failure cases myself when examining the code.

Probably an argument for using a result here for better debugging, but that should be done in a later PR.

**Comment by tormeh** [comment]
> Good! I'm not 100% convinced that this is exhaustive, but I couldn't find any other failure cases myself when examining the code.
> 
> Probably an argument for using a result here for better debugging, but that should be done in a later PR.

I don't have merge privileges, btw. Someone else will have to merge

**Comment by alice-i-cecile** [comment]
Yep, I'm one of the folks with merge rights. Once we have a second approval (from anyone), this PR will be marked with Ready-For-Final-Review and we'll merge it in :)

File: crates/bevy_render/src/camera/camera.rs
Comment (rparrett) [inline crates/bevy_render/src/camera/camera.rs:176]: I think this might not be entirely correct.

To me, it looks like this would return `None` only if there is no render target, or if the projection matrix / render target have not yet been computed by `camera_system`.

File: crates/bevy_render/src/camera/camera.rs
Comment (rparrett) [inline crates/bevy_render/src/camera/camera.rs:227]:
```suggestion
    /// Returns `None` if any of these conditions occur:
```

(And same change elsewhere)

File: crates/bevy_render/src/camera/camera.rs
Comment (Selene-Amanita) [reply crates/bevy_render/src/camera/camera.rs:176]: I agree with that, the "render target" cannot not be set, it's not an option, but the "computed render target",  `computed.target_info`, can not be set if the camera was just created or something.

Comment (rparrett) [reply crates/bevy_render/src/camera/camera.rs:176]: Ah, so I suppose it's just a matter of perspective. I am imagining reading this as a user with no knowledge of that internal implementation detail and it seems a bit confusing as-is from that angle.

As a user, I can set a custom viewport or not, and that doesn't really matter here.

As a user, the render target is not optional (but perhaps I can set it to something invalid?)

As a user, I might attempt to use these functions before `computed.whatever` is computed by `camera_system`.

I personally think that we should use this sort of user-facing language, or if we're talking about specific optional properties, maybe throw some backticks around them to make that more clear.

Comment (Selene-Amanita) [reply crates/bevy_render/src/camera/camera.rs:176]: I agree.

Actually, no matter if the viewport is set or not you need `Camera.computed.target_info`, so it doesn't matter at all if there is a viewport, and the user only needs to know that.

Something like "Returns `None` if the function is called just after the Camera is created, before `camera_system` is executed."

File: crates/bevy_render/src/camera/camera.rs
Comment (tormeh) [reply crates/bevy_render/src/camera/camera.rs]: Fixed

File: crates/bevy_render/src/camera/camera.rs
Comment (tormeh) [reply crates/bevy_render/src/camera/camera.rs]: Made a draft for this based on @Selene-Amanita's comment

Comment (Selene-Amanita) [reply crates/bevy_render/src/camera/camera.rs]: I investigated a bit more and edited my message, sorry @tormeh I saw that you already added my suggestion ^^'

Comment (Selene-Amanita) [reply crates/bevy_render/src/camera/camera.rs]: I think for `world_to_viewport` you can put a reference, like I think for `world_to_viewport` you can put a reference, like "The logical viewport size cannot be computed, see [logical_viewport_size](Camera::logical_viewport_size)"

File: crates/bevy_render/src/camera/camera.rs
Comment (Selene-Amanita) [inline crates/bevy_render/src/camera/camera.rs:238]: On this last point:

I think this reason should be the first in the least, it's the only "unexpected" one, the others are all like "something went wrong", this one is a design choice (to exclude stuff outside the frustum but only for front and back, not left/right or up/down). I think it kinda makes sense otherwise since we discard the z component we have no way to know if it's in the frustum or not, but it should be more "advertised".

Also, ***technically*** I feel like despite was is said in the comment bellow, and this is the case by default, if the "implicit frustum" defined by the projection matrix (the one used to actually render stuff), and the "actual frustum" (the one used to exclude stuff from the rendering process), are not the same (which, I think, can happen, but you'd have to do some weird stuff), this uses the "implicit frustum", not the actual one, that maybe should be mentioned?

File: crates/bevy_render/src/camera/camera.rs
Comment (Selene-Amanita) [inline crates/bevy_render/src/camera/camera.rs:367]: Same as https://github.com/bevyengine/bevy/pull/8841/files#r1246420948 except here if the Matrix is not set it would panic or put invalid values (depending on glam_assert) too because the 0 matrix can't be inverted.

The "quick fix" to return None early could be added for this function too.

File: crates/bevy_render/src/camera/camera.rs
Comment (tormeh) [reply crates/bevy_render/src/camera/camera.rs]: Done

File: crates/bevy_render/src/camera/camera.rs
Comment (tormeh) [reply crates/bevy_render/src/camera/camera.rs]: Pushed a potential fix. I want to restrict the scope of this PR to not involve any code changes, so haven't done any of that

Comment (Selene-Amanita) [reply crates/bevy_render/src/camera/camera.rs]: Fair ^^

We (you, me, someone else, idk) can make another PR later for that.

I'll review more thoroughly later but at a glance it seems good to me.

File: crates/bevy_render/src/camera/camera.rs
Comment (tormeh) [reply crates/bevy_render/src/camera/camera.rs]: Have also revised this

File: crates/bevy_render/src/camera/camera.rs
Comment (tormeh) [reply crates/bevy_render/src/camera/camera.rs:232]: I don't really understand this, but I've added a parentheses note

Comment (Selene-Amanita) [reply crates/bevy_render/src/camera/camera.rs:232]: I think you can change the line to "The computed coordinates are beyond the near or far plane defined by the projection,". Or maybe link to https://docs.rs/bevy/latest/bevy/render/camera/enum.Projection.html:

```rust
The computed coordinates are beyond the near or far plane defined by the [`Projection`],
```

I would also change the comment inside the code bellow to "(implicit, defined by the projection)"

For more details if you're interested, if you look at, for example, [Camera3dBundle](https://docs.rs/bevy/latest/bevy/core_pipeline/core_3d/struct.Camera3dBundle.html), you'll notice it has a [Projection](https://docs.rs/bevy/latest/bevy/render/camera/enum.Projection.html) component, and a [Frustum](https://docs.rs/bevy/latest/bevy/render/primitives/struct.Frustum.html) component.

The projection is used to transform all the objects of the world to fit them in a small box in front of the camera, this box will be "squished" to a rectangle that will make up the image rendered by the camera.

But because everything that won't end up in that box don't need to be transformed, we optimize beforehand which ones need to be. For there is the frustum, which is a kind of truncated pyramid that defines the limits of what the camera can see. We can check if an object is in the pyramid, if it isn't we don't consider it for rendering at all.

In almost all scenarios, the Frustum is calculated from the Projection, the truncated pyramid will become the rendering box after projection, so the near and far plane of both (where the pyramid is truncated for the Frustum, and the limits of the projection box) are the same. But in some specific scenarios they can not be, hence why I think it's important to specify it explicitly here.

File: crates/bevy_render/src/camera/camera.rs
Comment (Selene-Amanita) [inline crates/bevy_render/src/camera/camera.rs:348]:
```suggestion
    /// Returns `None` if the `camera_transform`, the `world_position`, or the projection matrix defined by [`CameraProjection`] contain `NAN`,
    /// or the projection matrix cannot be inverted.
    /// Panics if the projection matrix cannot be inverted and `glam_assert` is enabled.
```

**Comment by Selene-Amanita** [comment]
After https://github.com/bevyengine/bevy/pull/8841#discussion_r1250827639 and https://github.com/bevyengine/bevy/pull/8841#discussion_r1250898765 are handled this PR looks good to me.

It needs to be followed by another PR to handle the case where the projection Matrix isn't defined yet, though.

**Comment by tormeh** [comment]
> After [#8841 (comment)](https://github.com/bevyengine/bevy/pull/8841#discussion_r1250827639) and [#8841 (comment)](https://github.com/bevyengine/bevy/pull/8841#discussion_r1250898765) are handled this PR looks good to me.

Done. Also rebased on main. I suppose you squash it on merge?

**Review by Selene-Amanita — APPROVED** [review]
Looks good to me!

@tormeh Just so you know, the reason CI fails is because there is a new clippy lint in Rust 1.71.0, released today, it has nothing to do with your code, you need to wait for this PR to be reviewed and then merge: https://github.com/bevyengine/bevy/pull/9144

**Comment by tormeh** [comment]
> Looks good to me!
> 
> @tormeh Just so you know, the reason CI fails is because there is a new clippy lint in Rust 1.71.0, released today, it has nothing to do with your code, you need to wait for this PR to be merged and then merge to your branch: https://github.com/bevyengine/bevy/pull/9144

Wow, you guys are responsive! Anyway, yeah I figured. Thanks for telling me, though.

**Comment by alice-i-cecile** [comment]
Thanks @tormeh :D Yeah, this will get squashed on merge, which is very convenient for PR authors.

**Comment by mockersf** [comment]
This PR needs to be rebased before it can be merged

**Pull request merged**
